duchess-reflect/src/codegen.rs

```diff
@@ -134,238 +129,30 @@ impl ClassInfo {
             .map(|f| self.static_field_getter(f))
             .collect::<Result<_, _>>()?;
 
-        let upcast_impls = self.upcast_impls(upcasts)?;
-
-        let output = quote_spanned! {
-            self.span =>
-
-            #[allow(non_camel_case_types)]
-            pub struct #struct_name<#(#java_class_generics_with_defaults,)*> {
-                _empty: std::convert::Infallible,
-                _dummy: ::core::marker::PhantomData<(#(#java_class_generics,)*)>
+        let mro_tys = self.mro(upcasts)?;
+
+        let output = quote! {
+            duchess::plumbing::setup_class! {
+                struct_name: [#struct_name],
```

Comment:
just curious—why put all of them in `[]`? Does it simplify the resolution of the macro?

Reply:
It's not always necessary, but it is needed in some cases where a `,` is not a suitable "end delimiter", e.g. `$($t:tt)*` but also on occasion some other fragment specifiers. I used to only use it sometimes but I've wound up just adopting a consistent style of always including them.